src/utils/processUpload.js

```diff
@@ -1,8 +1,88 @@
+/* eslint-disable no-param-reassign */
+import { Storage } from 'aws-amplify';
+import pako from 'pako';
+import _ from 'lodash';
 import { createSample, updateSampleFile } from '../redux/actions/samples';
+import UploadStatus from './UploadStatus';
+
+const loadAndCompressIfNecessary = async (file, onLoaded, onError) => {
+  const inGzipFormat = file.bundle.mime === 'application/gzip';
+
+  const reader = new FileReader();
+  reader.onabort = () => onError();
+  reader.onerror = () => onError();
+  reader.onload = () => {
+    const loadedFile = reader.result;
+
+    if (inGzipFormat) {
+      onLoaded(loadedFile);
+    } else {
+      const compressed = pako.gzip(loadedFile, { to: 'string' });
+      onLoaded(compressed);
+    }
+  };
+
+  reader.readAsArrayBuffer(file.bundle);
+};
+
+const compressAndUpload = async (sample, activeProjectUuid, dispatch) => {
+  const updatedSampleFiles = Object.entries(sample.files).reduce((result, [fileName, file]) => {
+    const uncompressed = file.bundle.type !== 'application/gzip';
+
+    const newFileName = uncompressed ? `${fileName}.gz` : fileName;
+    const newFile = {
+      ...file,
+      name: newFileName,
+    };
+
+    result[newFileName] = newFile;
+
+    return result;
+  }, {});
+
+  Object.entries(updatedSampleFiles).map(async ([fileName, file]) => {
+    loadAndCompressIfNecessary(
+      file,
+      async (loadedFile) => {
+        const bucketKey = `${activeProjectUuid}/${sample.uuid}/${fileName}`;
+
+        return Storage.put(bucketKey, loadedFile)
+          .then(() => {
+            dispatch(updateSampleFile(sample.uuid, {
+              ...file,
+              status: UploadStatus.UPLOADED,
+            }));
+          })
+          .catch(() => {
+            dispatch(updateSampleFile(sample.uuid, {
+              ...file,
+              status: UploadStatus.UPLOAD_ERROR,
+            }));
+          });
+      },
+      async () => {
+        dispatch(updateSampleFile(sample.uuid, {
+          ...file,
+          status: UploadStatus.UPLOAD_ERROR,
+        }));
+      },
```

Comment:
These dispatched functions look like the `actions` functions that we usually implement, dispatching `actionTypes`. Any reason why you decide to implement it this way instead of using the common pattern?

Reply:
I'm not sure I understand the question, the `updateSampleFile` was already being used somewhere else too, it is a normal action function.

If the question is more about using the `UploadStatus` enum for the statuses I believe it is a clean way to standardize keys, which helps in making the code less prone to errors because instead of using the strings everywhere we use the constants storing them.

This way in the future we only need to modify the strings in the enum to change those keys for all code that depends on it without fear of breaking anything.

Reply:
Oh I understand now I think, you mean moving all this logic to an action, you're right that it doesn't fit that well with the logic followed elsewhere.

I guess since all of this was running from a function and not directly from a component I got kind of in a grey area in this respect, you're right that it makes sense to move all the logic of uploading and interacting with amplify to a particular dispatch.

Reply:
Yea, I mean instead of using `updateSampleFile` and passing in `status: UploadStatus`, would it be cleaner to have `UploadStatus`as an action type, and then create new action to handle that e.g. `fileUploaded`, `fileError`

Reply:
I mainly followed the logic in the whole processUpload file which does actually update other properties following this same pattern, so we would have to probably move most of this function over to redux if we wanted to follow a different pattern.

Furthermore, moving this to an action would involve moving the Storage.put() to an action too, which might be shifting part of this function's functionality over to a redux action instead of keeping it here.